38.md

```diff
@@ -0,0 +1,147 @@
+
+NIP-38
+======
+
+Encrypted Group Chat
+--------------------
+
+`draft` `optional` `author:vishalxl` 
+
+This NIP defines new event kinds for encrypted group chat. 
+
+It works by creating a shared secret that is used to encrypt the group or channel messages. Only the participants know that secret, so only such members can read/write to this group chat.
+
+It reserves three event kinds (140-142) for immediate use and 7 event kinds (143-149) for future use.
+
+- `140 - create encrypted channel`
+- `141 - change channel metadata including participants`
+- `142 - send encrypted message`
+
+
+## Kind 140: Create Encrypted channel
+
+Create a Encrypted chat channel.
+
+In the channel creation `content` field, Client SHOULD include basic channel metadata (`name`, `about` and `picture`).
+
+```json
+{
+    "content": "{\"name\": \"Demo Channel\", \"about\": \"A test channel.\", \"picture\": \"https://placekitten.com/200/200\"}",
+    ...
+}
+```
+
+The members of the group ( people who are allowed to post there) should be added as p tags to this event. The creator should also add themselves as p tag, if they want to participate there too.
+
+Further, the creator of the channel should send a Direct Message to each of the other participants, which will have the channel's shared-secret private/public keys. The Direct message should be sent by a kind other than 4; but presently kind 4 direct messages are used. 
+
+The format of that message is:
+
+`App Encrypted Channels: inviting you to encrypted channel <channel id> encrypted using private public keys <shared private key> <corresponding shared public key>`
+
+The three id's mentioned should each be 64 byte long hex encoded strings. No quotes are used.
+
+Clients that support this NIP will read that and know that they're invited to join the given encrypted channel, and they can use the given secret key to send/read messages in that channel.
+
+Other clients will decrypt this message, and ignore it, but the user can read that they have been invited to such a channel.
+
+Note: Clients for a user may create a channel where the only participant is the creator of the channel. Then new members can be added by sending a kind 141 message, as discussed further. 
+
+## Kind 141: Update metadata and Add participants
```

Comment:
Something me and @Semisol talked about awhile back was creating a new shared secret whenever a participant is removed, and this is encrypted into tags for each participant. 

The client would then query all the update messages and add all of the session keys to their current set. Then for older messages it would try to decrypt with each key in the set until a successful decryption worked.

This would enable you to re-key channels once participants are removed with a pretty simple algo. Also you don't need to send invites in this case either, the key is automatically encrypted to the user in the channel creation/update tags.

Reply:
Removing of participant ( or cases where a new secret is needed for all ) can be catered to by this method, where we allow the creator  (or anyone for that matter) to "fork" a group, so that the new fork will have a new id, new secret. We mark the new group as a 'fork' by adding some specific tag info. 

Then internally in clients, user can be seamlessly shown messages from what are two channels ( if you go by channel id itself), but are logically the same cause the client knows which channels are connected logically as same. 

This is the simplest way I think we can implement removal: it also has the logical consistency ( which will make coding easier imo), that each channel id is only ever connected with one shared secret. But using extra tags, we can establish a logical connection ( that 2+ channels are same channel as far as user knows).

Reply:
@paulmillr